File: src/main/kotlin/customfit/ai/kotlinclient/summaries/SummaryManager.kt
package customfit.ai.kotlinclient.summaries

import customfit.ai.kotlinclient.core.CFConfig
import customfit.ai.kotlinclient.core.CFUser
import customfit.ai.kotlinclient.network.HttpClient
import java.util.Collections
import java.util.Timer
import java.util.TimerTask
import java.util.concurrent.LinkedBlockingQueue
import java.util.concurrent.atomic.AtomicLong
import kotlin.concurrent.fixedRateTimer
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import mu.KotlinLogging
import org.joda.time.DateTime
import kotlinx.serialization.json.*
import kotlinx.serialization.encodeToString
import kotlinx.serialization.Serializable

private val logger = KotlinLogging.logger {}

// --- Copied Helper function to serialize Any --- 
private fun anyToJsonElement(value: Any?): JsonElement = when (value) {
    null -> JsonNull
    is JsonElement -> value // If it's already a JsonElement, return it directly
    is String -> JsonPrimitive(value)
    is Number -> JsonPrimitive(value)
    is Boolean -> JsonPrimitive(value)
    is Map<*, *> -> buildJsonObject { // Recursively handle maps
        value.forEach { (k, v) ->
            if (k is String) {
                put(k, anyToJsonElement(v)) // Recursive call
            } else {
                // Handle non-string keys if necessary, e.g., convert toString or throw error
                logger.warn { "Skipping non-string key in map during serialization: $k" }
            }
        }
    }
    is Iterable<*> -> buildJsonArray { // Recursively handle lists/collections
        value.forEach { 
            add(anyToJsonElement(it)) // Recursive call
        }
    }
    // Add other specific types if needed (e.g., Date -> JsonPrimitive(date.toString()))
    else -> throw kotlinx.serialization.SerializationException("Serializer for class '${value::class.simpleName}' is not found. Cannot serialize value of type Any.")
}
// --- End Helper function --- 

class SummaryManager(
        private val sessionId: String,
        private val httpClient: HttpClient,
        private val user: CFUser,
        private val cfConfig: CFConfig
) {
    // Use atomic values to allow thread-safe updates
    private val summariesQueueSize = cfConfig.summariesQueueSize
    private val summariesFlushTimeSeconds = cfConfig.summariesFlushTimeSeconds
    private val flushIntervalMs = AtomicLong(cfConfig.summariesFlushIntervalMs)
    
    private val summaries: LinkedBlockingQueue<CFConfigRequestSummary> = LinkedBlockingQueue(summariesQueueSize)
    private val summaryTrackMap = Collections.synchronizedMap(mutableMapOf<String, Boolean>())
    private val trackMutex = Mutex()
    private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())
    
    // Timer management
    private var flushTimer: Timer? = null
    private val timerMutex = Mutex()

    init {
        logger.info { "SummaryManager initialized with summariesQueueSize=$summariesQueueSize, summariesFlushTimeSeconds=$summariesFlushTimeSeconds, flushIntervalMs=${flushIntervalMs.get()}" }
        startPeriodicFlush()
    }

    /**
     * Updates the flush interval and restarts the timer
     * 
     * @param intervalMs new interval in milliseconds
     */
    suspend fun updateFlushInterval(intervalMs: Long) {
        require(intervalMs > 0) { "Interval must be greater than 0" }
        
        flushIntervalMs.set(intervalMs)
        restartPeriodicFlush()
        logger.info { "Updated summaries flush interval to $intervalMs ms" }
    }

    fun pushSummary(config: Any) {
        if (config !is Map<*, *>) {
            logger.warn { "Config is not a map: $config" }
            return
        }
        val configMap =
                config.takeIf { it.keys.all { k -> k is String } }?.let {
                    @Suppress("UNCHECKED_CAST")
                    (it as Map<String, Any>).also { 
                    }
                }
                        ?: run {
                            logger.warn { "Config map has non-string keys: $config" }
                            return
                        }

        val experienceId =
                configMap["experience_id"] as? String
                        ?: run {
                            logger.warn { "Missing mandatory 'experience_id' in config: $configMap" }
                            return
                        }
                        
        // Validate mandatory fields before creating the summary
        val configId = configMap["config_id"] as? String
        val variationId = configMap["variation_id"] as? String
        // Keep version as String but ensure it's not null
        val versionString = configMap["version"]?.toString()

        if (configId == null) {
            logger.warn { "Missing mandatory 'config_id' for summary: $configMap" }
            return
        }
        if (variationId == null) {
            logger.warn { "Missing mandatory 'variation_id' for summary: $configMap" }
            return
        }
        if (versionString == null) {
            logger.warn { "Missing or invalid mandatory 'version' for summary: $configMap" }
            return
        }

        scope.launch {
            trackMutex.withLock {
                if (summaryTrackMap.containsKey(experienceId)) {
                    logger.debug { "Experience already processed: $experienceId" }
                    return@launch
                }
                summaryTrackMap[experienceId] = true
            }

            val configSummary =
                    CFConfigRequestSummary(
                            config_id = configId,
                            version = versionString, // Use validated String version
                            user_id = configMap["user_id"] as? String,
                            requested_time = DateTime.now().toString("yyyy-MM-dd HH:mm:ss.SSSZ"), // Keep as formatted string for now
                            variation_id = variationId,
                            user_customer_id = user.user_customer_id,
                            session_id = sessionId,
                            behaviour_id = configMap["behaviour_id"] as? String,
                            experience_id = experienceId,
                            rule_id = configMap["rule_id"] as? String
                    )

            if (!summaries.offer(configSummary)) {
                logger.warn { "Summary queue full, forcing flush for: $configSummary" }
                flushSummaries()
                if (!summaries.offer(configSummary)) {
                    logger.error { "Failed to queue summary after flush: $configSummary" }
                }
            } else {
                logger.debug { "Summary added to queue: $configSummary" }
                // Check if queue size threshold is reached
                if (summaries.size >= summariesQueueSize) {
                    flushSummaries()
                }
            }
        }
    }

    suspend fun flushSummaries() {
        if (summaries.isEmpty()) {
            logger.debug { "No summaries to flush" }
            return
        }
        val summariesToFlush = mutableListOf<CFConfigRequestSummary>()
        summaries.drainTo(summariesToFlush)
        if (summariesToFlush.isNotEmpty()) {
            sendSummaryToServer(summariesToFlush)
            logger.info { "Flushed ${summariesToFlush.size} summaries successfully" }
        }
    }

    private suspend fun sendSummaryToServer(summaries: List<CFConfigRequestSummary>) {
        val jsonPayload =
                try {
                    // Use kotlinx.serialization builders
                    val jsonObject = buildJsonObject {
                        // Use user.toUserMap() and the helper
                        put("user", buildJsonObject { 
                            // Use helper function for values in user map
                            user.toUserMap().forEach { (k, v) -> 
                                put(k, anyToJsonElement(v))
                            } 
                        })
                        
                        // Add summaries array
                        put("summaries", buildJsonArray {
                            summaries.forEach { summary ->
                                // Assuming CFConfigRequestSummary is simple enough or @Serializable
                                // If not, it would need manual construction or its own serializer
                                add(Json.encodeToJsonElement(summary)) 
                                /* // Manual construction if needed:
                                add(buildJsonObject { 
                                    summary.config_id?.let { put("config_id", JsonPrimitive(it)) }
                                    summary.version?.let { put("version", JsonPrimitive(it)) }
                                    // ... other summary fields ...
                                })
                                */
                            }
                        })
                        
                        // Add SDK version
                        put("cf_client_sdk_version", JsonPrimitive("1.1.1")) // Use correct version
                    }
                    
                    Json.encodeToString(jsonObject)
                } catch (e: Exception) {
                     // Catch specific SerializationException from helper if needed
                    if (e is kotlinx.serialization.SerializationException) {
                        logger.error(e) { "Serialization error creating summary payload: ${e.message}" }
                    } else {
                        logger.error(e) { "Error serializing summaries: ${e.message}" }
                    }
                    summaries.forEach { this.summaries.offer(it) } // Re-queue on serialization error
                    return
                }

        // Print the API payload for debugging
        val timestamp = java.text.SimpleDateFormat("HH:mm:ss.SSS").format(java.util.Date())
        logger.debug { "================ SUMMARY API PAYLOAD ================" }
        logger.debug { jsonPayload }
        logger.debug { "====================================================" }

        val success =
                httpClient.postJson("https://api.customfit.ai/v1/config/request/summary?cfenc=${cfConfig.clientKey}", jsonPayload)
        if (!success) {
            logger.warn { "Failed to send ${summaries.size} summaries, re-queuing" }
            // Re-add summaries to queue in case of failure, but avoid infinite growth
            val capacity = summariesQueueSize - this.summaries.size
            if (capacity > 0) {
                summaries.take(capacity).forEach { this.summaries.offer(it) }
            }
        } else {
            logger.info { "Successfully sent ${summaries.size} summaries" }
        }
    }

    private fun startPeriodicFlush() {
        scope.launch {
            timerMutex.withLock {
                // Cancel existing timer if any
                flushTimer?.cancel()
                
                // Create a new timer
                flushTimer = fixedRateTimer("SummaryFlush", daemon = true, period = flushIntervalMs.get()) {
                    scope.launch {
                        logger.debug { "Periodic flush triggered for summaries" }
                        flushSummaries()
                    }
                }
            }
        }
    }
    
    private suspend fun restartPeriodicFlush() {
        timerMutex.withLock {
            // Cancel existing timer if any
            flushTimer?.cancel()
            
            // Create a new timer with updated interval
            flushTimer = fixedRateTimer("SummaryFlush", daemon = true, period = flushIntervalMs.get()) {
                scope.launch {
                    logger.debug { "Periodic flush triggered for summaries" }
                    flushSummaries()
                }
            }
            logger.debug { "Restarted periodic flush with interval ${flushIntervalMs.get()} ms" }
        }
    }
    
    // Method to retrieve all active summaries for other components
    fun getSummaries(): Map<String, Boolean> = summaryTrackMap.toMap()
}

File: customfit-reactnative-client-sdk/src/analytics/summary/SummaryManager.ts
import { CFResult } from '../../core/error/CFResult';
import { CFConfig } from '../../core/types/CFTypes';
import { HttpClient } from '../../network/HttpClient';
import { Storage } from '../../utils/Storage';
import { Logger } from '../../logging/Logger';
import { CFConstants } from '../../constants/CFConstants';
import { CFConfigRequestSummary, CFConfigRequestSummaryUtil } from './CFConfigRequestSummary';
import { ConnectionMonitor } from '../../platform/ConnectionMonitor';

/**
 * Summary manager for tracking config request summaries
 * Matches the functionality of Flutter and Kotlin SDKs
 */
export class SummaryManager {
  private readonly sessionId: string;
  private readonly httpClient: HttpClient;
  private currentUser: any;
  private readonly config: CFConfig;
  private readonly connectionMonitor: ConnectionMonitor;

  private readonly queueSize: number;
  private flushIntervalMs: number;
  private readonly flushTimeSeconds: number;

  private readonly summaries: CFConfigRequestSummary[] = [];
  private readonly trackMap: Map<string, boolean> = new Map();
  private flushTimer: NodeJS.Timeout | null = null;

  constructor(
    sessionId: string,
    httpClient: HttpClient,
    user: any,
    config: CFConfig
  ) {
    this.sessionId = sessionId;
    this.httpClient = httpClient;
    this.currentUser = user;
    this.config = config;
    this.connectionMonitor = ConnectionMonitor.getInstance();
    
    this.queueSize = config.summariesQueueSize || CFConstants.SummaryDefaults.QUEUE_SIZE;
    this.flushIntervalMs = config.summariesFlushIntervalMs || CFConstants.SummaryDefaults.FLUSH_INTERVAL_MS;
    this.flushTimeSeconds = config.summariesFlushTimeSeconds || CFConstants.SummaryDefaults.FLUSH_TIME_SECONDS;

    Logger.info(`📊 SummaryManager initialized with queueSize=${this.queueSize}, flushIntervalMs=${this.flushIntervalMs}, flushTimeSeconds=${this.flushTimeSeconds}`);
    this.startPeriodicFlush();
  }

  /**
   * Update the flush interval - matches Kotlin SDK
   */
  updateFlushInterval(intervalMs: number): CFResult<number> {
    try {
      if (intervalMs <= 0) {
        throw new Error('Interval must be greater than 0');
      }

      this.flushIntervalMs = intervalMs;
      this.restartPeriodicFlush();
      Logger.info(`📊 SUMMARY: Updated summaries flush interval to ${intervalMs} ms`);
      return CFResult.success(intervalMs);
    } catch (error) {
      Logger.error(`Failed to update flush interval to ${intervalMs}: ${error}`);
      return CFResult.errorFromException(error as Error);
    }
  }

  /**
   * Adds a configuration summary to the queue - matches Flutter SDK
   */
  async pushSummary(config: Record<string, any>): Promise<CFResult<boolean>> {
    // Log the config being processed
    Logger.info(`📊 SUMMARY: Processing summary for config: ${config.key || 'unknown'}`);

    // Validate input is a map
    if (typeof config !== 'object' || config === null) {
      const message = `Config is not an object: ${config}`;
      Logger.warning(`📊 SUMMARY: ${message}`);
      return CFResult.errorWithMessage(message);
  }

    // Validate required fields
    const experienceId = config.experience_id as string;
    if (!experienceId) {
      const message = 'Missing mandatory experience_id in config';
      Logger.warning(`📊 SUMMARY: ${message}, summary not tracked`);
      return CFResult.errorWithMessage(message);
    }

    const configId = config.config_id as string;
    const variationId = config.variation_id as string;
    const version = config.version?.toString();

    const missingFields: string[] = [];
    if (!configId) missingFields.push('config_id');
    if (!variationId) missingFields.push('variation_id');
    if (!version) missingFields.push('version');

    if (missingFields.length > 0) {
      const message = `Missing mandatory fields for summary: ${missingFields.join(', ')}`;
      Logger.warning(`📊 SUMMARY: ${message}, summary not tracked`);
      return CFResult.errorWithMessage(message);
    }

    // Prevent duplicates
    if (this.trackMap.has(experienceId)) {
      Logger.debug(`📊 SUMMARY: Experience already processed: ${experienceId}`);
      return CFResult.success(true);
    }

    this.trackMap.set(experienceId, true);

    const summary = CFConfigRequestSummaryUtil.fromConfig(
      config,
      this.currentUser?.userCustomerId || '',
      this.sessionId
    );

    Logger.info(`📊 SUMMARY: Created summary for experience: ${experienceId}, config: ${configId}`);

    // Check if queue is full
    if (this.summaries.length >= this.queueSize) {
      Logger.warning('📊 SUMMARY: Queue full, forcing flush for new entry');
      await this.flushSummaries();

      if (this.summaries.length >= this.queueSize) {
        Logger.error('📊 SUMMARY: Failed to queue summary after flush');
        return CFResult.errorWithMessage('Queue still full after flush');
      }
    }

    this.summaries.push(summary);
    Logger.info(`📊 SUMMARY: Added to queue: experience=${experienceId}, queue size=${this.summaries.length}`);

    // Check if queue size threshold is reached
    if (this.summaries.length >= this.queueSize) {
      Logger.info(`📊 SUMMARY: Queue size threshold reached (${this.summaries.length}/${this.queueSize}), triggering flush`);
      this.flushSummaries();
    }

    return CFResult.success(true);
  }

  /**
   * Flushes collected summaries to the server - matches Flutter/Kotlin SDK
   */
  async flushSummaries(): Promise<CFResult<number>> {
    if (this.summaries.length === 0) {
      Logger.debug('📊 SUMMARY: No summaries to flush');
      return CFResult.success(0);
    }

    const summariesToFlush = this.summaries.splice(0); // Drain the queue
    
    if (summariesToFlush.length === 0) {
      Logger.debug('📊 SUMMARY: No summaries to flush after drain');
      return CFResult.success(0);
    }

    Logger.info(`📊 SUMMARY: Flushing ${summariesToFlush.length} summaries to server`);

    try {
      const result = await this.sendSummaryToServer(summariesToFlush);
      if (result.isSuccess) {
        Logger.info(`📊 SUMMARY: Successfully flushed ${summariesToFlush.length} summaries to server`);
        return CFResult.success(summariesToFlush.length);
      } else {
        Logger.warning(`📊 SUMMARY: Failed to flush summaries: ${result.error?.message}`);
        return CFResult.errorWithMessage(`Failed to flush summaries: ${result.error?.message}`);
      }
    } catch (error) {
      Logger.error(`📊 SUMMARY: Unexpected error during summary flush: ${error}`);
      return CFResult.errorFromException(error as Error);
    }
  }

  /**
   * Create user payload matching the backend's CreateUserRequest DTO.
   */
  private createUserPayloadAlignedWithDTO(): Record<string, any> {
    const currentUser = this.currentUser as any; // Cast for type flexibility
    
    const properties: Record<string, any> = { ...(currentUser?.properties || {}) };

    // Add device to properties, attempting to match cURL/simplified structure first
    // If currentUser.device is richer and has toMap(), that could be used if backend expects it.
    if (currentUser?.device && typeof currentUser.device.toMap === 'function') {
        properties.device = currentUser.device.toMap(); 
    } else if (currentUser?.device) { // Simple object case
        properties.device = currentUser.device;
    } else {
        // Default device properties if nothing specific on currentUser
        // This matches the simplified structure from cURL, good for basic info
        properties.device = {
            os_name: "React Native", // Or Platform.OS
            sdk_type: "react-native",
            sdk_version: CFConstants.General.DEFAULT_SDK_VERSION
        };
  }
    // Ensure platform is in properties, as per cURL and common practice
    if (!properties.platform) {
        properties.platform = "React Native";
    }

    const userPayload: Record<string, any> = {
      anonymous: currentUser?.anonymous ?? false, // DTO default is false
      user_customer_id: currentUser?.userCustomerId || null,
      properties: properties,
      // hs_fields, private_fields, session_fields are omitted as per DTO defaults (empty/null)
      // and not typically part of basic summary user context.
    };

    // Clean null/undefined values from the top level of userPayload
    Object.keys(userPayload).forEach(key => {
      if (userPayload[key] === null || userPayload[key] === undefined) {
        delete userPayload[key];
  }
    });
    // Also clean properties specifically, as it's a map that DTO initializes to empty if not given
    if (userPayload.properties) {
        Object.keys(userPayload.properties).forEach(key => {
            if (userPayload.properties[key] === null || userPayload.properties[key] === undefined) {
                delete userPayload.properties[key];
            }
        });
        // If properties becomes empty after cleaning, remove it only if backend prefers absence
        // Given DTO defaults to new HashMap<>(), an empty {} should be fine.
    }

    return userPayload;
  }

  /**
   * Sends summary data to the server
   */
  private async sendSummaryToServer(summaries: CFConfigRequestSummary[]): Promise<CFResult<boolean>> {
    Logger.info(`📊 SUMMARY HTTP: Preparing to send ${summaries.length} summaries`);

    summaries.forEach((summary, index) => {
      Logger.debug(`📊 SUMMARY HTTP: Summary #${index + 1}: experience_id=${summary.experience_id}, config_id=${summary.config_id}, version=${summary.version}`);
    });

    const userMap = this.createUserPayloadAlignedWithDTO();
    const summariesMap = summaries.map(summary => CFConfigRequestSummaryUtil.toMap(summary));
    
    const payload = {
      user: userMap,
      summaries: summariesMap,
      cf_client_sdk_version: CFConstants.General.DEFAULT_SDK_VERSION,
    };

    // Log payload details for debugging
    Logger.debug(`📊 SUMMARY HTTP: User payload: ${JSON.stringify(userMap)}`);
    Logger.debug(`📊 SUMMARY HTTP: Summaries payload: ${JSON.stringify(summariesMap)}`);
    Logger.debug(`📊 SUMMARY HTTP: Full payload: ${JSON.stringify(payload)}`);

    // Use relative path since HttpClient already has the base URL
    const url = `/v1/config/request/summary?cfenc=${this.config.clientKey}`;

    Logger.info(`📊 SUMMARY: Sending ${summaries.length} summaries to server`);

    try {
      const result = await this.httpClient.post(url, payload);

      if (result.isSuccess && (result.data?.status === 200 || result.data?.status === 202)) {
        Logger.info(`📊 SUMMARY HTTP: Response code: ${result.data.status}`);
        Logger.info('📊 SUMMARY HTTP: Summary successfully sent to server');
        return CFResult.success(true);
      } else {
        const errorBody = result.data?.data || result.error?.message || 'No error body';
        Logger.warning(`📊 SUMMARY HTTP: Error code: ${result.data?.status || 'unknown'}`);
        Logger.warning(`📊 SUMMARY HTTP: Error body: ${errorBody}`);
        
        // Re-queue summaries on failure if we're offline
        if (!this.connectionMonitor.isConnected()) {
          this.handleSendFailure(summaries);
        }
        
        return CFResult.errorWithMessage(`API error response: ${result.data?.status || 'unknown'}`);
      }
    } catch (error) {
      Logger.error(`📊 SUMMARY HTTP: Exception: ${error}`);
      this.handleSendFailure(summaries);
      return CFResult.errorFromException(error as Error);
    }
  }

  /**
   * Helper method to handle send failures by re-queueing items
   */
  private handleSendFailure(summaries: CFConfigRequestSummary[]): void {
    Logger.warning(`📊 SUMMARY: Failed to send ${summaries.length} summaries after retries, re-queuing`);
    let requeueFailCount = 0;

    summaries.forEach(summary => {
      if (this.summaries.length >= this.queueSize) {
        requeueFailCount++;
      } else {
        this.summaries.push(summary);
      }
    });

    if (requeueFailCount > 0) {
      Logger.error(`📊 SUMMARY: Failed to re-queue ${requeueFailCount} summaries after send failure`);
  }
  }

  /**
   * Starts the periodic flush timer
   */
  private startPeriodicFlush(): void {
    try {
      // Cancel existing timer
    this.stopFlushTimer();
    
      // Create new timer
    this.flushTimer = setInterval(async () => {
        try {
          Logger.debug('📊 SUMMARY: Periodic flush triggered for summaries');
          await this.flushSummaries();
        } catch (error) {
          Logger.error(`📊 SUMMARY: Error during periodic summary flush: ${error}`);
        }
      }, this.flushIntervalMs);

      Logger.debug(`📊 SUMMARY: Started periodic summary flush with interval ${this.flushIntervalMs} ms`);
    } catch (error) {
      Logger.error(`📊 SUMMARY: Failed to start periodic summary flush: ${error}`);
    }
  }

  /**
   * Restarts the periodic flush timer with the current interval
   */
  private restartPeriodicFlush(): void {
    try {
      this.startPeriodicFlush();
      Logger.debug(`📊 SUMMARY: Restarted periodic flush with interval ${this.flushIntervalMs} ms`);
    } catch (error) {
      Logger.error(`📊 SUMMARY: Failed to restart periodic summary flush: ${error}`);
    }
  }

  /**
   * Stop the flush timer
   */
  private stopFlushTimer(): void {
    if (this.flushTimer) {
      clearInterval(this.flushTimer);
      this.flushTimer = null;
      Logger.debug('📊 SUMMARY: Flush timer stopped');
    }
  }

  /**
   * Returns all tracked summaries for debugging
   */
  getSummaries(): Map<string, boolean> {
    return new Map(this.trackMap);
        }

  /**
   * Get the current queue size
   */
  getQueueSize(): number {
    return this.summaries.length;
  }

  /**
   * Shutdown method to clean up timers
   */
  shutdown(): void {
    this.stopFlushTimer();
    Logger.info('📊 SummaryManager shutdown complete');
  }
} 